File: 0545-boundary-of-binary-tree/0545-boundary-of-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void formLeftBoundary(TreeNode node,List<Integer> res) {
        if(node==null) return;
        if(node.left!=null) {
            res.add(node.val);
            formLeftBoundary(node.left,res);
        }
        else if(node.right != null) {
            res.add(node.val);
            formLeftBoundary(node.right,res);
        }
    }
    
    public void formRightBoundary(TreeNode node,List<Integer> res) {
        if(node==null) return;
        if(node.right!=null) {
            formRightBoundary(node.right,res);
            res.add(node.val);
        }
        else if(node.left!=null) {
            formRightBoundary(node.left,res);
            res.add(node.val);
        }
    }
    public void addLeaves(TreeNode node,List<Integer> res) {
        if(node==null) return;
        addLeaves(node.left,res);
        if(node.left==null && node.right == null)res.add(node.val);
        addLeaves(node.right,res);
    }
    public List<Integer> boundaryOfBinaryTree(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if(root==null) return res;
        res.add(root.val);
        formLeftBoundary(root.left,res);
        addLeaves(root.left,res);
        addLeaves(root.right,res);
        formRightBoundary(root.right,res);
        return res;
    }
}

File: 0545-boundary-of-binary-tree/0545-boundary-of-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
     void formLeftBoundary(TreeNode* node,vector<int>& res) {
        if(!node) return;
        if(node->left) {
            res.push_back(node->val);
            formLeftBoundary(node->left,res);
        }
        else if(node->right) {
            res.push_back(node->val);
            formLeftBoundary(node->right,res);
        }
    }
    
     void formRightBoundary(TreeNode* node,vector<int>& res) {
        if(!node) return;
        if(node->right) {
            formRightBoundary(node->right,res);
            res.push_back(node->val);
        }
        else if(node->left) {
            formRightBoundary(node->left,res);
            res.push_back(node->val);
        }
    }
    void addLeaves(TreeNode* node,vector<int>& res) {
        if(!node) return;
        addLeaves(node->left,res);
        if(!node->left && !node->right)res.push_back(node->val);
        addLeaves(node->right,res);
    }
    vector<int> boundaryOfBinaryTree(TreeNode* root) {
        vector<int> res;
        if(!root) return res;
        res.push_back(root->val);
        formLeftBoundary(root->left,res);
        addLeaves(root->left,res);
        addLeaves(root->right,res);
        formRightBoundary(root->right,res);
        return res;
    }
};